### DISCUSSION TITLE
imports in generated files are not conform with "verbatimModuleSyntax" ?

### DISCUSSION THREAD

**Discussion by TimRensen** [odata2ts/odata2ts#277] [Q&A] [answered]
Hello everyone,

currently I am trying to implement odata2ts to my new project. I am overall pretty new to TypeScript and Generators, and therefore my question might be solved easily by configuration. 

I setup a new vue-project with typescript and unit tests (vitest). Because vue is using the "verbatimModuleSyntax" in the compilerOptions of the tsconfig.json, building the project seems impossible with a type-checker.

While the project is working completely fine during development, building throws following errors during the tests:

`error TS1484: '...' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.`

This problem is also highlighted in the generated QObjects and Services:
![image](https://github.com/odata2ts/odata2ts/assets/37070855/a4051bf5-0c14-4378-b06c-23d978083240)
![image](https://github.com/odata2ts/odata2ts/assets/37070855/c7d7ad5b-852d-423f-bdb9-8664b6691b34)

Out of curiosity I changed the imports to "import type" in the QObjects and Service files and the problem seems to be gone, I was able to build the project without errors.
![image](https://github.com/odata2ts/odata2ts/assets/37070855/e4ba785f-d760-428f-ba68-b656717e3ce3)
![image](https://github.com/odata2ts/odata2ts/assets/37070855/a41229b4-742f-4045-941f-805162f86e49)

Because I have to generate the QObjects and Services every once in a while, for API changes, adding the "type" manually can't be a valid solution. 

I tried excluding the generated files for the vitest but this seems to create other errors.

Do you have any ideas, hints or solutions for my problem? Is it possible to generate the imports with the "type-only import" for those files?

I am open for any ideas and welcome to any comments. Thanks for taking your time!

Greetings!

**Comment by texttechne** [comment]
Hi @TimRensen,
please open an issue for that:: Should make sense to always add the `type` modifier for type only imports.

As workaround: 
* create a seperate tsconfig.json and switch `verbatimModuleSyntax` off (extending should make this easy)
* switch to `emitMode: EmitModes.js_dts` 
* specify the new tsconfig.json for that: See {[option `tsconfig` of the Base Settings](https://odata2ts.github.io/docs/generator/configuration#base-settings)

Hope this helps.

**Comment by TimRensen** [comment]
Hi @texttechne,

thanks for the quick response! I transferred the problem into an issue.

For now I exclude the generated files from the type-check (vitest) so they won't throw an error while testing and building. 
Thank you!

**Comment by texttechne** [comment, marked as answer]
Hey @TimRensen & @brease-colin,

the type specifier will be added in a future release, but I regard it as a minor enhancement.
The "workaround" I've described above, really should be considered the standard use case. Well, at least in my mind... But I would like your opinion though. 

I think the generated code should be considered as library, like any other library from `node_modules`. As such you should prefer to emt `js` and `d.ts` files (having the freedom to specify an own `tsonfig.json`). Also, you don't want to type check the generated code. As with any other library you assume that the library code is working like specified 😁

In the next release I want to go as far as prefixing each generated file with `@ts-nocheck` to prevent any type checking whatsoever. This idea came up in the context of really huge OData services where type checking may run out of memory and just takes a considerable amount of time in any case. Only turning on `debug: true` would then allow to type check the generate code. But it would be an exception (debugging odata2ts), not the regular use case.

So does this perspective resonate with you? And most importantly: Do you have any compelling reason to emit TypeScript instead?

**Reply by TimRensen** [reply]
Hey @texttechne,

I totally agree that it would be a minor enhancement and ignoring the generated files for type checking should be the way to go. That's the way I implemented it currently after your workaround message.
I dont have any specific reason to emit TypeScript instead of `js `and `d.ts`. During my initial implementation of the package I got it running easier with the TypeScript files.

**Reply by brease-colin** [reply]
Sounds fine to me as well, thanks for the quick and extensive response!
